traffic_ops/traffic_ops_golang/deliveryservice/letsencryptcert.go

```diff
@@ -0,0 +1,305 @@
+	response, err := tx.Exec(q, fqdn, value)
+	tx.Commit()
+	if err != nil {
+		log.Errorf("Inserting dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+		return errors.New("Inserting dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+	} else {
+		rows, err := response.RowsAffected()
+		if err != nil {
+			log.Errorf("Determining rows affected dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+			return errors.New("Determining rows affected dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+		}
+		if rows == 0 {
+			log.Errorf("Zero rows affected when inserting dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+			return errors.New("Zero rows affected when inserting dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+		}
+	}
+
+	return nil
+}
+
+func (d *DNSProviderTrafficRouter) CleanUp(domain, token, keyAuth string) error {
+	fqdn, value := dns01.GetRecord(domain, keyAuth)
+	tx, err := d.db.Begin()
+
+	q := `DELETE FROM dnschallenges WHERE fqdn = $1 and record = $2`
+	response, err := tx.Exec(q, fqdn, value)
+	tx.Commit()
+	if err != nil {
+		log.Errorf("Deleting dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+		return errors.New("Deleting dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+	} else {
+		rows, err := response.RowsAffected()
+		if err != nil {
+			log.Errorf("Determining rows affected when deleting dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+			return errors.New("Determining rows affected when deleting dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+		}
+		if rows == 0 {
+			log.Errorf("Zero rows affected when deleting dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+			return errors.New("Zero rows affected when deleting dns txt record for fqdn '" + fqdn + "' record '" + value + "': " + err.Error())
+		}
+	}
+
+	return nil
+}
+
+func GenerateLetsEncryptCertificates(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	ctx, _ := context.WithTimeout(r.Context(), LetsEncryptTimeout)
+
+	req := tc.DeliveryServiceLetsEncryptSSLKeysReq{}
+	if err := api.Parse(r.Body, nil, &req); err != nil {
+		api.HandleErr(w, r, nil, http.StatusBadRequest, errors.New("parsing request: "+err.Error()), nil)
+		return
+	}
+	if *req.DeliveryService == "" {
+		req.DeliveryService = req.Key
+	}
+
+	go GetLetsEncryptCertificates(inf.Config, req, ctx, inf.User)
+
+	api.WriteRespAlert(w, r, tc.WarnLevel, "Beginning async call to Let's Encrypt for "+*req.DeliveryService+".  This may take a few minutes.")
```

Comment:
This seems more like an "info" level alert to me, tbh.

Also, if you'd like, this is a good use case for [the `202 Accepted` HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/202).

Reply:
done